File: chrome-ext/background.js
chrome.runtime.onInstalled.addListener(() => {
  console.log('running background script');
  chrome.contextMenus.create({
    "id": "cnotes",
    "title": "Lookup with Chinese Notes ...",
    "contexts": ["selection"]
  });
});

chrome.contextMenus.onClicked.addListener((info) => {
  console.log(`CNotes context menu clicked ${info.selectionText}`);
  chrome.tabs.query({active: true, currentWindow: true}, (tabs) => {
    const tabId = tabs[0].id;
    console.log(`sendMessage to tab ${tabId}`)
    chrome.tabs.sendMessage(tabId, {term: info.selectionText});
  });
});

File: web/background.dart
// Entry point for Firefox add-on.
// Firefix add-ons do not support service workers, so use a background script
// instead.

import 'dart:js';

void contextMenuSetup() {
  var jsOnClicked = context['browser']['contextMenus']['onClicked'];
  JsObject onClicked = (jsOnClicked is JsObject
      ? jsOnClicked
      : new JsObject.fromBrowserObject(jsOnClicked));
  onClicked.callMethod('addListener', [onMenuEvent]);
}

void onMenuEvent(JsObject info, var tabsNotUsed) async {
  var query = info['selectionText'];
  print('onMenuEvent query: ${query}');
}

void setUpApp() async {
  var contextMenuText = 'Lookup with Chinese Notes ...';
  var menuObj = JsObject.jsify({
    'id': 'cnotes',
    'title': contextMenuText,
    'contexts': ['selection']
  });
  context['browser']['contextMenus'].callMethod('create', [menuObj]);
}

void main() async {
  print('cnotes background enter');
  setUpApp();
  contextMenuSetup();
}
